xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,246 @@
+// next evaluation      XXXXXX║ 0   ║ 1   ║ 2   ║ 3   ║ 4   ║ 5   ║ 6   ║ ...
+//                      ──────╨─────╨─────╨─────╨─────╨─────╨─────╨─────╨────
+// do_recv                    ┌─────┐     ┌─────────────────┐           ┌────
+//                      ──────┘     └─────┘                 └───────────┘    
+//                      ──────╥─────╥─────╥─────╥─────╥─────╥────────────────
+// symbol, count        XXXXXX║ A,2 ║XXXXX║ B,1 ║ B,1 ║ C,3 ║XXXXXXXXXXXXXXXX
+// (input channel)      ──────╨─────╨─────╨─────╨─────╨─────╨────────────────
+// last                                   ┌─────┐     ┌─────┐                
+// (input channel)      ──────────────────┘     └─────┘     └────────────────
+//                      ╥─────╥───────────╥───────────╥─────────────────╥────
+// state.symbol         ║ 0   ║ A         ║ B         ║ C               ║ 0  
+// (set state value)    ╨─────╨───────────╨───────────╨─────────────────╨────
+//                      ╥─────╥─────╥─────╥───────────╥─────╥─────╥─────╥────
+// state.count          ║ 0   ║ 1   ║ 0   ║ 0         ║ 2   ║ 1   ║ 0   ║ 0  
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//                                                               
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐    
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_interface as rle_interface
+
+type DecInData  = rle_interface::EncData;
+type DecOutData = rle_interface::PreData;
+
+// structure to preserve the state of an RLE decoder
+struct RLEDecState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    // symbol to be repeated on output
+    symbol: bits[SYMBOL_WIDTH],
+    // count of symbols that has to be send
+    count: bits[COUNT_WIDTH],
+    // send last when repeat ends
+    last: bool,
+}
+ // RLE decoder implementation
+pub proc RLEDec<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+    init {(
+        RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH> {
+            symbol: bits[SYMBOL_WIDTH]:0,
+            count:  bits[COUNT_WIDTH]:0,
+            last:   bool:false,
+        }
+    )}
+
+    config (
+        input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+        output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+    ) {(input_r, output_s)}
+
+    next (tok: token, state: RLEDecState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+        let zero_input = DecInData { symbol: bits[SYMBOL_WIDTH]:0, count: bits[COUNT_WIDTH]:0, last: false };
+        let empty = state.count == bits[COUNT_WIDTH]:0;
```

Comment:
should we rename `empty` to `symbol_completed` to be more explicit?

Reply:
I think empty was good, but this will be more descriptive, so I'll change it